README.md
# xpcba

xpcba（暂称）是万莫斯统一测试套件

## 缘起

为什么我们需要一套新的测试工具？目前我们在产测时使用多种测试工具，在开发和维护上需花费不少时间成本，且会随着平台的增多而增加。另外目前所使用的工具拓展性不佳。

## 优点

1. 多接口支持，可使用 adb 、eth 或 wifi 连接 设备端 与 客户端；
2. 多平台支持，客户端可使用 Windows 或 Linux 的 PC，或者 ROCK Pi；
3. 设备系统要求低，设备端可运行任意系统，只需支持 Python 3 即可；
4. 工具拓展性良好，可给测试任务添加钩子，如录音回传播放，如上传测试日志；
5. 测试程序要求低，可使用 C 或 Shell 或 Python 编写测试程序，只需符合约定即可；
6. 测试任务易分配，每个测试项相互独立，测试任务可分配给多人编写；
7. 测试程序易复用，使用配置文件描述板子，修改配置文件可复用部分程序；
8. 客户端可 Hack，测试时可修改设备端程序，此可用于小修改而不用重烧录固件；

## 缺点

1. 在最初编写测试程序需花费不少时间；
2. 尚没用于大规模生产中，可行性待验证；
3. 工具处于试验中，可能会有 bug；
4. 熟悉新工具需花费时间；

## 进度

1. 基本设计已完成，可使用命令行测试，GUI 还没编写；
2. 上传日志服务端还没开始编写；

## 运行

1. 设备端

   拷贝 core 目录至设备，运行 server.py

   ```bash
   $ python3 core/server.py
   ```

2. 客户端

   ```bash
   $ adb forward tcp:6688 tcp:6688
   $ python3 core/client.py  # 测试交互是否成功
   ```

   目前提供了 ui/cli.py，可作为测试的入口

   ```bash
   $ python3 ui/cli.py -b vc098
   ```

   此会读取 `conf/vc098.json` 的测试项，进入测试并打印测试结果。

## 生成测试镜像

1. 正常生成Debian CLI镜像，需要tmux
2. 运行以下命令：

   ````bash
   #!/bin/bash

   # On host: apt install multipath-tools qemu-user-static

   SOURCE_IMG="radxa-zero2_debian_bullseye_cli.img"
   # To view test result via monitor
   #TEST_ENV="getty@tty1"
   # or via serial
   # Need to update `ttyAML0` to real serial device
   TEST_ENV="serial-getty@ttyAML0"
   PRODUCT="rs102"

   # Make a backup
   cp "$SOURCE_IMG" xpcba.img

   # Mount image
   ROOT_DEV=$(sudo kpartx xpcba.img | tail -n 1 | cut -d ' ' -f 1)
   sudo kpartx -a xpcba.img
   mkdir ./mnt
   sudo mount /dev/mapper/$ROOT_DEV ./mnt

   # Set up auto login
   SYSTEMD_OVERRIDE=./mnt/etc/systemd/system/$TEST_ENV.service.d
   sudo mkdir -p $SYSTEMD_OVERRIDE
   cat << EOF | sudo tee $SYSTEMD_OVERRIDE/override.conf
   [Service]
   ExecStart=
   EOF
   if grep -q "serial" <<< $TEST_ENV
   then
       AUTOLOGIN="ExecStart=-/sbin/agetty --autologin root -o '-p -- \\\\u' --noclear %I \$TERM"
   else
       AUTOLOGIN="ExecStart=-/sbin/agetty --autologin root -o '-p -- \\\\u' --keep-baud 115200,57600,38400,9600 %I \$TERM"
   fi
   sudo tee -a $SYSTEMD_OVERRIDE/override.conf <<< $AUTOLOGIN

   # Download xpcba
   sudo git clone --depth 1 -b master http://gitlab.vamrs.com/xpcba/xpcba.git ./mnt/xpcba

   # Set up xpcba auto start
   cat << EOF | sudo tee -a ./mnt/root/.bashrc
   dmesg -n 1
   tmux new-session -d python3 /xpcba/core/server.py
   tmux new-session python3 /xpcba/ui/cli.py --board $PRODUCT --sku d4e16 --via local
   EOF

   # Delete root password so auto login won't ask for password
   sudo chroot ./mnt/ passwd --delete root

   # Clean up
   sudo umount ./mnt
   rm -rf ./mnt
   sudo kpartx -d xpcba.img
   ````

## 约定

1. 配置文件

   ui/config.json 为全局配置文件，conf 存放项目配置文件，如 vc098.json；

2. 测试程序

   bin 存放可运行文件或测试脚本，以项目名为文件夹组织。

   <del> 为了兼容 rkpcba，工具沿用了rkpcba 的交互格式，</del>程序在测试完成时需打印以下格式信息，

      ```text
      <msg>,<PASS|FAIL>,<ERR_CODE>
      ```

   如：

      ```text
      <bt_test>,<PASS>,<0>
      <ddr_test: size:256MB>,<FAIL>,<-54>
      ```

3. 参考程序

   可参考 `misc/rs309.sh` 去编写测试时的临时脚本，和对比它正式运行时的 `conf/rs309.json` 配置文件。


changelog.rst
Changelog
=========

vNext - TBD
^^^^^^^^^^^^^^^^^^^^

**Added**

- Add generic I2C test, currently used in RS102 (Redmine #1648)

v0.0.5 - 2021-12-25
^^^^^^^^^^^^^^^^^^^^

**Added**

- core/utils/Timer
- ui/cli: status bar
- source env file
- condition test item

**Fixed**

- core/utils/enter_test
- core/utils/get_path

**Changes**

- core/server: add date to logfile
- core/protocol: increase packet size
- conf/json: use bool true/false
- conf/json: add node field

v0.0.4 - 2021-08-18
^^^^^^^^^^^^^^^^^^^^

**Changes**

- modify project structure

v0.0.3 - 2021-07-01
^^^^^^^^^^^^^^^^^^^^

**Added**

- plugin: iperf3
- boards/rc115

**Changes**

- modify pull params
- modify push params

v0.0.2 - 2021-06-19
^^^^^^^^^^^^^^^^^^^^

**Added**

- protocol

**Breaking Changes**

- Not currently compatible with RKPCBA

v0.0.1 - 2021-06-13
^^^^^^^^^^^^^^^^^^^^

Initial version

**Added**

- server
- client 
- client-cli 
- boards/vc098


ui/cli.py
#!/usr/bin/python3
import argparse
import json
import os
import sys
import time
from multiprocessing import freeze_support

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from core.client import Client, __version__, __author__
from core.utils import _exec, adb_forward, enter_test, fmt_ret, get_path, get_env, Timer


class FakeTimer:
    def __init__(self, *args, **kw):
        pass

    def __getattr__(self, name):
        return lambda *a, **k: None


def merge(arr, k, v):
    for x in arr:
        if x['exec'] == k:
            x.update(v)


def conf(args):
    with open(get_path(args.config)) as f:
        for k, v in json.loads(f.read()).items():
            if not args.__dict__.get(k):
                args.__dict__[k] = v
        args.task = args.task or args.board

    with open(get_path(f'conf/{args.board}/{args.task}.json')) as f:
        config = json.loads(f.read())
        args.update = config.get('update', 'undefined')

    if args.env and config.get('source'):
        args.__dict__.update(get_env(config['source']))

    if args.sku:
        with open(get_path(f'conf/{args.board}/{args.sku}.json')) as f:
            for k, v in json.loads(f.read()).items():
                merge(config['items'], k, v)

    if args.debug:
        globals()['Timer'] = FakeTimer

    items, conds = [], []
    for item in config['items']:
        if len(item['name']) > 13:
            raise Exception(f'The name is too long.\n  {item["name"]}')

        if item.get('pass'):
            continue

        d = {
            'name': item['name'],
            'args': {
                'cmd': item['exec'],
                'type': item.get('type', 'buildin'),
                'node': item.get('node', 'bin')
            },
            'cond': item.get('cond'),
            'hint': item.get('hint'),
            'rest': item.get('rest', 0),
            'auto': item.get('auto', True),
            'wait': item.get('wait', False)
        }

        for key in ('args', 'save', 'hook'):
            if item.get(key) is not None:
                d['args'][key] = item[key]

        if item.get('type') == 'external':
            d['args']['file'] = config['scripts'][item['exec']]

        if item.get('cond'):
            conds.append(d)
        else:
            items.append(d)

    return items, conds, config.get('finish')


def print_res(idx, res, item):
    ret = fmt_ret(res['RESULT'], item['auto'])
    err, msg = res['ERR_CODE'], res['MSG']
    print(f'  {idx:>2}. {item["name"]:<15}{ret}{err:<4}{msg}')


def verify(cond, result):
    match = {
        'all:pass': lambda x: not any(x),
        'any:pass': lambda x: not all(x),
        'all:fail': lambda x: all(x),
        'any:fail': lambda x: any(x)
    }

    def some(cond, result):
        key, ret = cond.split(':')
        if key not in result.keys():
            raise Exception('Not a valid condition.')
        return bool(result[key]) == bool(ret == 'fail')

    if cond in match.keys():
        ret = match[cond](result.values())
    else:
        ret = some(cond, result)
    return ret


def main(args):
    print(f'XPCBA v{__version__} client starting up ...')
    # print(f'Powered by {__author__}')

    items, conds, finish = conf(args)
    query, result, cmdlen = {}, {}, 0

    if args.via == 'adb':
        adb_forward(args.port, args.ser, args.timeout, args.debug)

    client = Client(args.host, args.port, args.debug)
    enter_test(client, args.board, args.timeout, args.cache)

    with open(get_path('env.json', client.dtmp), 'w') as f:
        f.write(json.dumps(args.__dict__))

    with open(get_path('items.json', client.dtmp), 'w') as f:
        f.write(json.dumps(items + conds))

    print(f'- board: {args.board}, sku: {args.sku or "default"}')
    print(f'- update: {args.update}')
    print('======== Start Test =========')

    t = Timer(600, wait=1.0, func='spin')
    for item in items:
        name, args = item['name'], item['args']
        cmdlen = max(cmdlen, len(args['cmd']))
        t.puts(item['hint'] or 'wait ack')
        print(f'  {name}')
        client.start(args)
        query[name] = 1
        if item.get('wait') == True:
            while True:
                time.sleep(0.5)
                if client.query(args).get('RESULT'):
                    break
        time.sleep(item['rest'])
        t.stop(f'\r{" " * 80}', '\r')

    print('======== Test Result ========')
    time.sleep(0.5)

    t = Timer(600, wait=0.05, func='spin')
    while True:
        for item in items:
            name, args = item['name'], item['args']
            if query[name]:
                res = client.query(args)
                if res.get('RESULT'):
                    query[name] = 0
                    idx = len(query) - sum(query.values())
                    result[args['cmd']] = int(res['ERR_CODE'])
                    t.stop(f'\r{" " * 80}', '\r')
                    print_res(idx, res, item)
                    time.sleep(0.1)
                else:
                    t.puts(f'query {res["TEST_ITEM"]:<{cmdlen}}')
                    time.sleep(1.05)

        if not any(query.values()):
            break

    with open(get_path('res.json', client.dtmp), 'w') as f:
        f.write(json.dumps(result))

    if conds:
        t = Timer(600, wait=0.05, func='spin')
        print('======== In The End =========')
        for idx, item in enumerate(conds, 1):
            name, args = item['name'], item['args']
            if verify(item['cond'], result):
                t.puts(item['hint'] or f'  Exec {name}')
                client.start(args)
                time.sleep(0.5)
                while True:
                    try:
                        res = client.query(args)
                    except Exception:
                        if 'reset key' in name.lower():  # for power key
                            res = {
                                'MSG': '',
                                'RESULT': 'PASS',
                                'ERR_CODE': '0'
                            }
                    if res.get('RESULT'):
                        t.stop(f'\r{" " * 80}', '\r')
                        print_res(idx, res, item)
                        break
                    time.sleep(1.05)
            else:
                print(f"# Cond '{item['cond']}' not met, skip {name}.")

    if finish:
        _exec(get_path(finish), client.dtmp)

    try:
        client.exit()
    except Exception as ex:
        if client.debug:
            print(ex)


if __name__ == '__main__':
    freeze_support()

    parser = argparse.ArgumentParser()
    parser.add_argument('-b', '--board')
    parser.add_argument('-v', '--via')
    parser.add_argument('-s', '--sku')
    parser.add_argument('-t', '--task')
    parser.add_argument('-S', '--ser', default='')
    parser.add_argument('-c', '--config', default='ui/config.json')
    parser.add_argument('-C', '--cache', action='store_true')
    parser.add_argument('-d', '--debug', action='store_true')
    parser.add_argument('-e', '--env', action='store_true')
    parser.add_argument('-H', '--host', default='127.0.0.1')
    parser.add_argument('-p', '--port', default=6688, type=int)
    parser.add_argument('-T', '--timeout', default=30, type=int)
    args = parser.parse_args()

    main(args)

    input('\nPress enter key to exit\n')
    time.sleep(180)
